docs: add presentation on refactoring (#13)
outlier-org/onboarding: refactoring.md

3. [feat: show review page if past lock date](https://github.com/outlier-org/calculus-static/pull/5699)

The above example is a good example, where I was refactoring to understand about the assignment.

By the time, that I completed the refactoring, there's was enough understanding of the code and the feature was completed with a small change

### 4.3 Litter-Pickup Refactoring

- I understand how the code works but realize that it's doing it badly
- It's not exactly necessary to refactor
- If it takes small effort, do it right away
- If it takes longer to fix it, make a note and fix it later. work on the feature instead

[refactor(WritingGradeCenter): move LoadingSpinner inside SubmissionTable](https://github.com/outlier-org/admin-static/pull/1234)

This PR above is the type of refactoring which we can do when we see something wrong in the codebase.

We could just fix the issue given that the solution doesn't take much time.
Otherwise creating a separate ticket is advised.

### 4.4 Planned Refactoring

- Set aside a time to refactor a piece of code
- Not related to any features
- Need to simplify code that's been giving trouble to the team regularly

> 
> [Refactor navigateToUuid function in Navigation Component](https://github.com/outlier-org/calculus-static/issues/2306)
> 
> Here, we create a plan and dedicate a certain amount of time focusing solely on improving a particular piece of code.

Reviewer: ```suggestion

This is an example [PR](https://github.com/outlier-org/calculus-static/issues/2306) for which certain amount of time was dedicated to solely improve a particular piece of code.
```

### 4.5 Long-Term Refactoring

- Some refactoring work can take a team weeks to complete
- Instead of dedicated/planned refactoring, gradually work on simplifying the code one small chunk at a time.

> [Refactor NavigationComponent](https://github.com/outlier-org/calculus-static/issues/5077)
> 
> This is a type of refactoring where we know we will need significant amount of time to get the work done.
> 
> Here, people all over the team get to try new things and improve the codebase a small step at a time.

Reviewer: ```suggestion
This [project] (https://github.com/outlier-org/calculus-static/issues/5077) serves as an example for one of those times when significant amount of time is needed to get the work done. In such instances, the collaboration of many people is needed to improve the code a small step at a time.```

## 5. When Should I Not Refactor?

- If code is messy but it works and it doesn't need to be modified, it doesn't need to be refactored. You can treat it as a library.
  - For e.g. [wiris](https://github.com/outlier-org/calculus-static/blob/master/public/quizzes/resources/mathtype.js) [scripts](https://github.com/outlier-org/calculus-static/blob/master/public/quizzes/resources/quizzes.js) in calculus-static
- If it is small enough that it could be rewritten without much effort.
  - For e.g. components like [CompletedMarker](https://github.com/outlier-org/calculus-static/blob/master/src/Components/CompletedMarker/CompletedMarker.jsx) in calculus-static

## 6. Two Hats

This is a method for working on any feature, not just refactoring

- While working on any feature, divide the time between two distinct activities: adding functionality and refactoring.
- We should only focus on one thing at a time.
- When adding functionality, don't think about the structure of the code.
- When refactoring, don't change any functionality in the code.
- https://twitter.com/KentBeck/status/1111050408007360512

> Again a good example of this approach would be the one we discussed on "Preparatory Refactoring"

Reviewer: ```suggestion
Again a good example of this approach would be the one discussed under "Preparatory Refactoring"
```
